Surgically add the EmperorViewController class

ElDeveloper opened the pull request:
After starting the implementation of `AnimationsViewController`, I noticed that there would be a fair amount of repeated code between this new class, `AxesController` and `EmperorAttributeABC`, therefore I decided to move that redundant code to a new base class `EmperorViewController`.

This new class understands what a decomposition view dictionary is, but doesn't really alter the functionality that EmperorViewControllerABC provides. The tests and most of the code were extracted from `EmperorAttributeABC`.

antgonza commented:
I think it looks fine 👍 . @ElDeveloper do we need another reviewer?

ElDeveloper commented:
Another reviewer would be OK with me, but perhaps not necessary.

On (Jul-28-16| 7:01), Antonio Gonzalez wrote:
> I think it looks fine 👍 . @ElDeveloper do we need another reviewer? 
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/528#issuecomment-235903256

antgonza commented on emperor/support_files/js/view-controller.js:
```diff
@@ -253,6 +239,75 @@ define([
+   * @param {String} k Key corresponding to active decomposition view.
+   */
+  EmperorViewController.prototype.setActiveDecompViewKey = function(k) {
+    // FIXME: this should be validated against decompViewDict i.e. we should be
```
Just noticed this, should this be an issue?

wasade replied:
I can look through in a moment

On Jul 28, 2016 08:29, "Antonio Gonzalez" <notifications@github.com> wrote:

> In emperor/support_files/js/view-controller.js
> <https://github.com/biocore/emperor/pull/528#discussion_r72644778>:
>
> > +  /**
> > +   * Retrieves the metadata field currently being controlled
> > +   *
> > +   * @return {String} A key corresponding to the active decomposition view.
> > +   */
> > +  EmperorViewController.prototype.getActiveDecompViewKey = function() {
> > +    return this.activeViewKey;
> > +  };
> > +
> > +  /**
> > +   * Changes the metadata column name to control.
> > +   *
> > +   * @param {String} k Key corresponding to active decomposition view.
> > +   */
> > +  EmperorViewController.prototype.setActiveDecompViewKey = function(k) {
> > +    // FIXME: this should be validated against decompViewDict i.e. we should be
>
> Just noticed this, should this be an issue?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/emperor/pull/528/files/6b118f7f6dd9154c17ff1dfb19d2ad7b90b32b97#r72644778>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAc8si9SNDo1TteFVUWBKcwmcbxM2SQFks5qaMrJgaJpZM4JWvVz>
> .
>

wasade replied:
sorry, this comment got placed in the wrong place, meant i can look through as a second reviewer

wasade commented on emperor/support_files/js/axes-controller.js:
```diff
@@ -8,7 +8,7 @@ define([
 ], function($, _, DecompositionView, ViewControllers, d3, contextmenu) {
 
   // we only use the base attribute class, no need to get the base class
```
is this comment still accurate?

ElDeveloper replied:
You are right, this is out of date.

On (Jul-28-16| 8:34), Daniel McDonald wrote:
> > @@ -8,7 +8,7 @@ define([
> >  ], function($, _, DecompositionView, ViewControllers, d3, contextmenu) {
> >  
> >    // we only use the base attribute class, no need to get the base class
> 
> is this comment still accurate?
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/528/files/6b118f7f6dd9154c17ff1dfb19d2ad7b90b32b97#r72645917

wasade commented on tests/javascript_tests/test_view_controller.js:
```diff
@@ -133,6 +134,117 @@ requirejs([
+     * Tests to make sure the exceptions are being raised as expected
+     *
+     */
+    test('Cosnstructor test exceptions', function(assert) {
```
Constructor

wasade commented:
Minor comments, :+1: otherwise

mortonjt commented on emperor/support_files/js/view-controller.js:
```diff
@@ -253,6 +239,75 @@ define([
+
+  /**
+   *
+   * @class EmperorViewControllerABC
```
Update?

mortonjt commented:
Solid.  👍

ElDeveloper commented:
Thanks all for the review. All comments should be addressed. @antgonza,
I didn't open a new issue, I just fixed the problem :) and added a test
to check the exception is raised. Aside from that all the other comments
were also addressed.

On (Jul-28-16| 9:21), Jamie Morton wrote:
> Solid.  👍
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/528#issuecomment-235947147

ElDeveloper commented:
Thanks everyone!

On (Jul-28-16|13:00), Antonio Gonzalez wrote:
> Merged #528.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/528#event-738363313
